<a href="https://colab.research.google.com/github/BrianMoringa/Week-3-IP-Submission-Brian-Onchweri-/blob/main/Copy_of_Moringa_Data_Science_Prep_W3_Independent_Project_2022_07_Brian_Onchweri.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Importing the Libraries to be used 

In [ ]:
# Let's import the pandas library
#
# OUR CODE GOES HERE
import pandas as pd
# as well as the Numpy library
import numpy as np

# OUR CODE GOES HERE

## Loading our Datasets

In [ ]:
#Loading the cells_geo description DataFrame

# We create a URL to Excel file (alternatively this can be a filepath)
url = '/content/cells_geo_description.xlsx'

# Then load the first sheet of the Excel file into a data frame
df = pd.read_excel(url, sheet_name=0, header=1)

# Lastly, view the first ten rows
df.head(10)

VILLES                                               City  String
0        STATUS                                  In Service or not  String
1  LOCALISATION                                  in ABIDJAN or not  String
2    DECOUPZONE                                  Geographical Zone  String
3      ZONENAME                                       Name of Zone  String
4     LONGITUDE                                          Longitude   Float
5      LATITUDE                                           Latitude   Float
6        REGION                                             Region  String
7          AREA                                               Area  String
8       CELL_ID                                     ID of the cell  String
9     SITE_CODE  Site (there are several cells per site, severa...  String

In [ ]:
df.describe()

VILLES               City  String
count       10                 10      10
unique      10                 10       2
top     STATUS  In Service or not  String
freq         1                  1       8

In [ ]:
#Loading the Cells Geo DataFrame


path = '/content/cells_geo.csv'

df1 = pd.read_csv('/content/cells_geo.csv')

# Lastly, view the first ten rows
df1.head(10)

;VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE
0  0;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
1  1;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
2  2;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
3  3;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
4  4;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
5  5;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
6  6;KRIKOREA;In Service;INTERIEUR;Grand-OUEST;"K...                                              
7  7;ABOBO;In Service;ABIDJAN;"""Abidjan_NORD";NP...                                              
8  8;ABOBO;In Service;ABIDJAN;"""Abidjan_NORD";NP...                                              
9  9;ABOBO;;ABIDJAN;"""Abidjan_NORD";NPONON;-4.02...

In [ ]:
#We're Checking and dropping duplicates if any exist 
df1.duplicated()
df1.drop_duplicates()

;VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE
0     0;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
1     1;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
2     2;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
3     3;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
4     4;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
...                                                 ...                                              
3969  3969;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              
3970  3970;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              
3971  3971;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              
3972  3972;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              
3973  3973;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              

[3974 rows x 1 columns]

In [ ]:
#checking for NuLL values and Dropping them subsequently
df1.isnull()
df1.dropna()

;VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE
0     0;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
1     1;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
2     2;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
3     3;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
4     4;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
...                                                 ...                                              
3969  3969;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              
3970  3970;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              
3971  3971;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              
3972  3972;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              
3973  3973;ZUENOULA;In Service;INTERIEUR;Grand-NORD;...                                              

[3974 rows x 1 columns]

In [ ]:

df1.describe()


;VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE
count                                                3974                                              
unique                                               3974                                              
top     0;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
freq                                                    1

**OUR TABLE ABOVE IS POORLY INDEXED, WE WILL CREATE A NEW TABLE AND POPULATE IT WITH VALUES FROM OUR ORIGINAL TABLE**

In [ ]:
#creating columns for our table
Columns=['STATUS','LOCALISATION','DECOUPZONE','ZONENAME','LONGITUDE','LATITUDE','REGION','AREA','CELL_ID','SITE_CODE']
#creating new DataFrame

new_df=pd.DataFrame(df1, columns=Columns,)

In [ ]:
#previewing what our table would look like
new_df


STATUS  LOCALISATION  DECOUPZONE  ...  AREA  CELL_ID  SITE_CODE
0        NaN           NaN         NaN  ...   NaN      NaN        NaN
1        NaN           NaN         NaN  ...   NaN      NaN        NaN
2        NaN           NaN         NaN  ...   NaN      NaN        NaN
3        NaN           NaN         NaN  ...   NaN      NaN        NaN
4        NaN           NaN         NaN  ...   NaN      NaN        NaN
...      ...           ...         ...  ...   ...      ...        ...
3969     NaN           NaN         NaN  ...   NaN      NaN        NaN
3970     NaN           NaN         NaN  ...   NaN      NaN        NaN
3971     NaN           NaN         NaN  ...   NaN      NaN        NaN
3972     NaN           NaN         NaN  ...   NaN      NaN        NaN
3973     NaN           NaN         NaN  ...   NaN      NaN        NaN

[3974 rows x 10 columns]

**WE CAN NOW GO AHEAD AND POPULATE OUR TABLE WITH VALUES USING FOR LOOPS**

In [ ]:
from pandas.core.arrays.categorical import factorize_from_iterable
#creating a for loop
#for index, row in df1.iterrows():


In [ ]:

# Loading CDR_description DataFrame
# 
# We create a URL to Excel file (alternatively this can be a filepath)
path = '/content/CDR_description.xlsx'

# Then load the first sheet of the Excel file into a data frame
df2 = pd.read_excel(path, sheet_name=0, header=1)

# Lastly, view the first ten rows
df2.head(10)

PRODUCT                                       Voice or SMS   String
0            VALUE                                      Billing price  Integer
1        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
2     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
3  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
4  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
5        COUNTRY_A                  Country of party A (useless here)   String
6        COUNTRY_B                  Country of party B (useless here)   String
7          CELL_ID                                     ID of the cell   String
8          SITE_ID                                     ID of the SITE   String

In [ ]:
#checking NuLL values 
df2.isnull()

PRODUCT  Voice or SMS  String
0    False         False   False
1    False         False   False
2    False         False   False
3    False         False   False
4    False         False   False
5    False         False   False
6    False         False   False
7    False         False   False
8    False         False   False

In [ ]:
df2.info

<bound method DataFrame.info of            PRODUCT                                       Voice or SMS   String
0            VALUE                                      Billing price  Integer
1        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
2     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
3  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
4  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
5        COUNTRY_A                  Country of party A (useless here)   String
6        COUNTRY_B                  Country of party B (useless here)   String
7          CELL_ID                                     ID of the cell   String
8          SITE_ID                                     ID of the SITE   String>

In [ ]:
df2.describe() 


PRODUCT   Voice or SMS  String
count        9              9       9
unique       9              9       2
top      VALUE  Billing price  String
freq         1              1       7

In [ ]:

#Loading the Telecom_dataset 1 DataFrame

# We create a URL to Excel file (alternatively this can be a filepath)
path = '/content/Telcom_dataset.csv'

df3 = pd.read_csv('/content/Telcom_dataset.csv')

# Lastly, view the first ten rows
df3.head(10)


PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
5    data      0  2012-05-06 23:07:26.0  ...     OTHER  /lrUXVGVcN  /lrUXVGVcN
6   Voice      0  2012-05-06 23:08:02.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
7   Voice      0  2012-05-06 23:08:03.0  ...     19e2e  09UjmmY1Ds  09UjmmY1Ds
8   Voice      0  2012-05-06 23:08:05.0  ...     19e2e  0DeimCpZOc  0DeimCpZOc
9     sms      0  2012-05-06 23:08:43.0  ...     19e2e  01MDiY5F89  01MDiY5F89

[10 rows x 10 columns]

In [ ]:
#checking NuLL values 
df3.isnull()

PRODUTC  VALUE  DATETIME  ...  COUNTRY_B  CELL_ID  SITE_ID
0       False  False     False  ...      False    False    False
1       False  False     False  ...      False    False    False
2       False  False     False  ...      False    False    False
3       False  False     False  ...      False    False    False
4       False  False     False  ...      False    False    False
...       ...    ...       ...  ...        ...      ...      ...
4996    False  False     False  ...      False    False    False
4997    False  False     False  ...      False    False    False
4998    False  False     False  ...      False    False    False
4999    False  False     False  ...      False    False    False
5000    False  False     False  ...      False    False    False

[5001 rows x 10 columns]

In [ ]:
df3.info

<bound method DataFrame.info of      PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2        sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
...      ...    ...                    ...  ...       ...         ...         ...
4996     sms    100  2012-05-07 00:01:54.0  ...     OTHER  0c2371d9bc  1e751d08a4
4997   Voice      1  2012-05-07 00:01:54.0  ...     19e2e  5fb791cf56  65efe2bc62
4998   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  7773627c12  9ef9616350
4999   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  3f4af49d6c  969b8d0e39
5000   Voice      7  2012-05-07 00:01:54.0  ...     19e2e  6f4d9b7

In [ ]:
df3.describe()

VALUE
count  5001.000000
mean     25.869626
std     101.934502
min       0.000000
25%       0.000000
50%       0.000000
75%      25.000000
max    4440.000000

In [ ]:

#Loading the Telecom_Dataset2 DataFrame

# We create a URL to Excel file (alternatively this can be a filepath)
path = '/content/Telcom_dataset2.csv'

df4 = pd.read_csv('/content/Telcom_dataset2.csv')

# Lastly, view the first ten rows
df4.head(10)


PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO
5     sms      0  2012-05-07 23:05:41.0  ...     19e2e  /MMplBlWjh  /MMplBlWjh
6   Voice     41  2012-05-07 23:06:14.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
7   Voice     11  2012-05-07 23:06:28.0  ...     19e2e  +gySf02tKW  +gySf02tKW
8     sms      0  2012-05-07 23:06:35.0  ...     OTHER  /0FxP3az1x  /0FxP3az1x
9     sms      0  2012-05-07 23:06:50.0  ...     19e2e  0ecjmwPX21  0ecjmwPX21

[10 rows x 10 columns]

In [ ]:
df4.describe()

VALUE
count  5001.000000
mean     24.334933
std      97.471586
min       0.000000
25%       0.000000
50%       0.000000
75%      25.000000
max    3380.000000

In [ ]:

#Loading the Telecom_Dataset3 DataFrame

# We create a URL to Excel file (alternatively this can be a filepath)
path = '/content/Telcom_dataset3.csv'

df5 = pd.read_csv('/content/Telcom_dataset3.csv')

# Lastly, view the first ten rows
df5.head(10)

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA
5   Voice     10  2012-05-08 23:03:59.0  ...     19e2e  +e+xcimwre  +e+xcimwre
6     sms      0  2012-05-08 23:06:06.0  ...     19e2e  +R5ccmtyJf  +R5ccmtyJf
7   Voice     10  2012-05-08 23:06:55.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
8     sms     25  2012-05-08 23:07:09.0  ...     19e2e  13O1+JEun3  13O1+JEun3
9     sms      0  2012-05-08 23:07:38.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[10 rows x 10 columns]

In [ ]:
df5.describe()

VALUE
count   5001.000000
mean      26.441512
std      224.200217
min        0.000000
25%        0.000000
50%        0.000000
75%       25.000000
max    12900.000000

In [ ]:
df5.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4996    False
4997    False
4998     True
4999    False
5000    False
Length: 5001, dtype: bool

In [ ]:
df5.info

<bound method DataFrame.info of      PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0      Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1        sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2      Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3        sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA
...      ...    ...                    ...  ...       ...         ...         ...
4996   Voice      0  2012-05-09 00:01:43.0  ...     19e2e  2aa04dcab7  f7d63a29e5
4997   Voice     62  2012-05-09 00:01:43.0  ...     19e2e  7f1f2f889f  b4ddb9e93c
4998   Voice     32  2012-05-09 00:01:43.0  ...     19e2e  a4731e780e  d68260ef50
4999   Voice     30  2012-05-09 00:01:43.0  ...     19e2e  98ed726da2  b4130b54ab
5000   Voice     60  2012-05-09 00:01:43.0  ...     19e2e  1c11a2f

#Concutinating df3, df4 and df5 into one DataFrame Called CombinedDataFrame

In [ ]:
CombinedDataFrame = pd.concat([df3, df4, df5], ignore_index=True, sort=True)
CombinedDataFrame

CELLID     CELL_ID CELL_ON_SITE  ...     SIET_ID     SITE_ID VALUE
0             NaN  /TJNe+Mmtu            1  ...         NaN  /TJNe+Mmtu     0
1             NaN  /+cKZKIp41            0  ...         NaN  /+cKZKIp41     0
2             NaN  /u0FSD+ahi            0  ...         NaN  /u0FSD+ahi    25
3             NaN  +cNeJzsTp3            1  ...         NaN  +cNeJzsTp3    56
4             NaN  051KhYDCpv            1  ...         NaN  051KhYDCpv    14
...           ...         ...          ...  ...         ...         ...   ...
14998  2aa04dcab7         NaN            3  ...  f7d63a29e5         NaN     0
14999  7f1f2f889f         NaN            1  ...  b4ddb9e93c         NaN    62
15000  a4731e780e         NaN            1  ...  d68260ef50         NaN    32
15001  98ed726da2         NaN            2  ...  b4130b54ab         NaN    30
15002  1c11a2fcbb         NaN            2  ...  6d70dae094         NaN    60

[15003 rows x 16 columns]

**FINDING MOST USED CITY FOR 3 DAYS**

In [ ]:
CombinedDataFrame.isna()


CELLID  CELL_ID  CELL_ON_SITE  ...  SIET_ID  SITE_ID  VALUE
0        True    False         False  ...     True    False  False
1        True    False         False  ...     True    False  False
2        True    False         False  ...     True    False  False
3        True    False         False  ...     True    False  False
4        True    False         False  ...     True    False  False
...       ...      ...           ...  ...      ...      ...    ...
14998   False     True         False  ...    False     True  False
14999   False     True         False  ...    False     True  False
15000   False     True         False  ...    False     True  False
15001   False     True         False  ...    False     True  False
15002   False     True         False  ...    False     True  False

[15003 rows x 16 columns]

In [ ]:
#WE CAN SEE WE HAVE NAN VALUES LETS DROP THEM
CombinedDataFrame.dropna()

Empty DataFrame
Columns: [CELLID, CELL_ID, CELL_ON_SITE, COUNTRY_A, COUNTRY_B, DATETIME, DATE_TIME, DW_A_NUMBER, DW_A_NUMBER_INT, DW_B_NUMBER, DW_B_NUMBER_INT, PRODUCT, PRODUTC, SIET_ID, SITE_ID, VALUE]
Index: []

In [ ]:
CombinedDataFrame.sort_values(by="VALUE",ascending=0)

CELLID     CELL_ID CELL_ON_SITE  ...     SIET_ID     SITE_ID  VALUE
11552  2def69183e         NaN            0  ...         NaN         NaN  12900
11838  3f5ac97214         NaN            2  ...  a64fc672f1         NaN   6750
1228          NaN  ffa6759bb2            _  ...         NaN         NaN   4440
10460  MpOhP3OKNN         NaN            2  ...  MpOhP3OKNN         NaN   4212
6072          NaN  5a7f6cf0fa            2  ...         NaN  2e828e607c   3380
...           ...         ...          ...  ...         ...         ...    ...
7257          NaN  ffa6759bb2            _  ...         NaN         NaN      0
7258          NaN  ffa6759bb2            _  ...         NaN         NaN      0
1164          NaN  N7Sq+rBs2r            2  ...         NaN  N7Sq+rBs2r      0
1163          NaN  ffa6759bb2            _  ...         NaN         NaN      0
0             NaN  /TJNe+Mmtu            1  ...         NaN  /TJNe+Mmtu      0

[15003 rows x 16 columns]

THE CITY ON INDEX **11552** IS THE MOST USED